### PR TITLE
Add unrolling to the legalization passes

### PR DESCRIPTION
**Pull request by s-perron**
Adds unrolling to the legalization passes.

After enabling unrolling I found a bug when there is a self-referencing
phi node.  That has been fixed.

The test that checks for that the order of optimizations is correct also
needed to be updated.

### CODE REVIEW

File: source/opt/loop_unroller.cpp [hunk 1]
```diff
@@ -842,19 +851,21 @@ void LoopUnrollerUtilsImpl::AssignNewResultIds(BasicBlock* basic_block) {
   }
 }
 
-// For all instructions in |basic_block| check if the operands used are from a
-// copied instruction and if so swap out the operand for the copy of it.
-void LoopUnrollerUtilsImpl::RemapOperands(BasicBlock* basic_block) {
-  for (Instruction& inst : *basic_block) {
-    auto remap_operands_to_new_ids = [this](uint32_t* id) {
-      auto itr = state_.new_inst.find(*id);
+void LoopUnrollerUtilsImpl::RemapOperands(Instruction* inst) {
```
Comment (dnovillo) [inline source/opt/loop_unroller.cpp]: Can you keep a variant of the original comment?  The new version doesn't have any documentation.

File: source/opt/loop_unroller.cpp [hunk 2]
```diff
@@ -842,19 +851,21 @@ void LoopUnrollerUtilsImpl::AssignNewResultIds(BasicBlock* basic_block) {
...
+      *id = itr->second;
+    }
+  };
+
+  inst->ForEachInId(remap_operands_to_new_ids);
+}
 
-    inst.ForEachInId(remap_operands_to_new_ids);
+void LoopUnrollerUtilsImpl::RemapOperands(BasicBlock* basic_block) {
+  for (Instruction& inst : *basic_block) {
```
Comment (dnovillo) [inline source/opt/loop_unroller.cpp]: Or just move the original comment here.  It seems to still apply.

File: source/opt/loop_unroller.cpp (hunk 1 above)
Comment (s-perron) [reply source/opt/loop_unroller.cpp]: The code standards that we use want the comments describing the function just before the declaration, not the definition.  The declaration at line 255 already has the comments, so the ones I removed are redundant.

File: source/opt/loop_unroller.cpp (hunk 2 above)
Comment (dneto0) [reply source/opt/loop_unroller.cpp]: Yah, generally the method comment describes behaviour from the perspective of the caller.  Comments about the algorithm's internal operation (that are not relevant to the interface) belong inside the function body.

**Comment by dnovillo** [comment]
Yeah. I keep forgetting that this is not optimized for my own workflow. I'm
a heavy tags user, tags take you to the definition not the declaration.

**Pull request merged**